File: fallout.py
#!/usr/bin/python
# helps hack terminals in Fallout 3 and 4
# copyright 2015, Styopa Semenukha, GPL v3+
from __future__ import print_function
import sys, re

def hamming(op1, op2):
    i = 0
    for c1, c2 in zip(op1, op2):
        if c1 == c2:
            i += 1
    return i

passwords = map(lambda p: p.lower(), sys.argv[1:])

if len(passwords) < 1:
    print("Too few arguments: %i" % len(passwords), file = sys.stderr)
    print("USAGE: %s PASSWORD PASSWORD..." % sys.argv[0])
    exit(1)

len_first = len(passwords[0])
for password in passwords:
    if len(password) != len_first:
        print("Passwords are of different length: '%s' and '%s'"
                % (passwords[0], password), file = sys.stderr)
        exit(2)

for each in passwords[:]:
    line = raw_input("Enter password and number of correct letters\n> ")
    (entered, correct) = tuple(re.split('\W+', line))
    entered = entered.lower()
    correct = int(correct)
    if (entered not in passwords):
        print("%s is not in the list of possible passwords" % entered, file = sys.stderr)
        exit(3)
    passwords.remove(entered)
    for password in passwords:
        if hamming(password, entered) != correct:
            passwords.remove(password)

    num = len(passwords)
    if num > 1:
        print("Possible passwords: " + ' '.join(passwords))
    elif num == 1:
        print("The password is: " + passwords[0])
        exit(0)
    else:
        print("No more passwords left", file = sys.stderr)
        exit(4)

File: fallout.pl
#!/usr/bin/perl
# helps hack terminals in Fallout 3
# copyright 2014, Styopa Semenukha, GPL v3+
use strict;
use warnings;
use diagnostics;
$\ = "\n";

if ($#ARGV < 1) {
	my $num = $#ARGV + 1;
	die "Too little arguments: $num\n" .
		"USAGE: $0 PASSWORD PASSWORD..."
}

my $len = length($ARGV[0]);
foreach my $arg (@ARGV) {
	if ( length($arg) != $len ) {
		die 'Passwords are of different length: ' .
			"'${ARGV[0]}' and '$arg'";
	}
}

sub hamming {
	my $op1 = shift; # ref to array
	my @op2 = split(//, shift);
	my $i = 0;
	for (0 .. $#op2) {
		$i++ if $$op1[$_] eq $op2[$_];
	}
	return $i;
}

# make 'em lowercase and convert to hash keys
my %passwords = map {$_ => 0} map(lc, @ARGV);

my ($passwd, $correct);
for (0 .. $#ARGV) {
	{
		local $\ = '';
		print "Enter password and number of correct letters\n> ";
	}
	$_ = <STDIN>;
	exit unless defined;
	($passwd, $correct) = split /\W+/;
	$passwd = lc($passwd);
	$correct = int($correct);
	unless (exists $passwords{$passwd}) {
		die "$passwd is not in the list of possible passwords";
	}
	my @entered = split(//, $passwd);
	delete $passwords{$passwd};
	foreach (keys %passwords) {
		unless ( hamming(\@entered, $_) == $correct ) {
			delete $passwords{$_};
		}
	}
	my $num = keys %passwords;
	if ($num > 1) {
		print "Possible passwords: " . join(' ', keys %passwords);
	} elsif ($num = 1) {
		print 'The password is: ' . join('', keys %passwords);
		exit;
	} else {
		die 'No more passwords left';
	}
}
